libs/context/BridgeContext.tsx

```diff
@@ -113,58 +128,88 @@ export function BridgeProvider({ children }: PropsWithChildren) {
 		setGasToken(bridgeTokenInput.token as TrnToken);
 	}, [bridgeTokenInput.token, network, setGasToken]);
 
-	const { trnApi } = useTrnApi();
-	const { userSession } = useWallets();
-	const authenticationMethod = useAuthenticationMethod();
-
-	const futurepass = userSession?.futurepass as string | undefined;
-
-	const { estimateFee, submitExtrinsic, xamanData } = useExtrinsic({
-		extrinsic: state.tx as sdk.Extrinsic,
-		senderAddress: futurepass,
-		feeOptions: {
-			assetId: state.gasToken.assetId,
-			slippage: +state.slippage / 100,
-		},
-	});
-
-	useEffect(() => {
-		if (network !== "root" || !state.tx) return;
-
-		estimateFee()
-			.then((gasFee) => setEstimatedFee(new Balance(gasFee, state.gasToken).toHuman()))
-			.catch(({ cause }: Error) => {
-				if (!cause) return;
-
-				updateState({
-					gasToken: DEFAULT_GAS_TOKEN,
-				});
-			});
-	}, [state.tx, state.gasToken, estimateFee, network]);
-
 	const buildTransaction = useCallback(
-		({ token, amount, toAddress }: { token?: Token; amount?: string; toAddress?: string }) => {
-			if (!token || !amount || !toAddress) return updateState({ tx: undefined });
+		async ({
+			token,
+			amount,
+			toAddress,
+		}: {
+			token?: Token;
+			amount?: string;
+			toAddress?: string;
+		}) => {
+			if (!token || !amount || !toAddress || !signer || !userSession)
+				return updateState({ builder: undefined, tx: undefined });
 
 			if (network === "root") {
-				if (!trnApi) return;
+				if (!trnApi || !customEx) return;
 
 				const bridgeToken = token as TrnToken;
-
 				const bridgeBalance = new Balance(amount, bridgeToken, false);
-
 				const decoded = decodeAccountID(toAddress.trim());
-
 				const decodedToAddress = `0x${convertStringToHex(decoded as any)}`;
 
-				const tx = trnApi.tx.xrplBridge.withdraw(
+				let tx = trnApi.tx.xrplBridge.withdraw(
 					bridgeToken.assetId,
 					bridgeBalance.toPlanckString(),
 					decodedToAddress,
 					null
 				);
 
-				updateState({ tx });
+				let builder = await createBuilder(
+					userSession,
+					state.gasToken.assetId,
+					state.slippage,
+					customEx,
+					tx
+				);
+
+				const { gasString, gasFee } = await builder.getGasFees();
+				const [gas] = gasString.split(" ");
+				setEstimatedFee(gas);
+
+				const gasTokenBalance = await builder.checkBalance({
+					walletAddress: userSession.futurepass,
+					assetId: state.gasToken.assetId,
+				});
+				const gasBalance = new Balance(+gasTokenBalance.balance, gasTokenBalance)
+					.toUnit()
+					.toNumber();
+
+				let canPay: boolean | undefined;
+				let amountWithoutGas: Balance<TrnToken> = bridgeBalance;
+				if (bridgeToken.assetId === state.gasToken.assetId) {
+					amountWithoutGas = bridgeBalance.toPlanck().minus(+gasFee * 1.5); // Safety margin for gas
+					canPay = amountWithoutGas.toUnit().toNumber() >= 0 ? true : false; // TODO 768
```

Comment:
```suggestion
					canPay = amountWithoutGas.toUnit().toNumber(); // TODO 768
```

What does the TODO comment here reference?